main.js

```diff
@@ -0,0 +1,171 @@
+                });
+                break;
+
+                case "caps": 
+                keyElement.classList.add("keyboard__key--wide", "keyboard__key--activatable");
+                keyElement.innerHTML = createIconHTML("keyboard_capslock");
+
+                keyElement.addEventListener("click", () => {
+                    this._toggleCapsLock();
+                    keyElement.classList.toggle("keyboard__key--active"/*, this.properties.capsLock*/); 
+                });
+                break;
+
+                case "enter": 
+                keyElement.classList.add("keyboard__key--wide");
+                keyElement.innerHTML = createIconHTML("keyboard_return");
+
+                keyElement.addEventListener("click", () => {
+                    this.properties.value += "\n";
+                    this._triggerEvent("oninput");
+                });
+                break;
+
+                case "space": 
+                keyElement.classList.add("keyboard__key--extra--wide");
+                keyElement.innerHTML = createIconHTML("space_bar");
+
+                keyElement.addEventListener("click", () => {
+                    this.properties.value += " ";
+                    this._triggerEvent("oninput");
+                });
+                break;
+
+                case "done": 
+                keyElement.classList.add("keyboard__key--wide", "keyboard__key--dark");
+                keyElement.innerHTML = createIconHTML("check_circle");
+
+                keyElement.addEventListener("click", () => {
+                    this.close();
+                    this._triggerEvent("onclose");
+                });
+                break;
+
+                default: 
+                keyElement.textContent = key.toLowerCase();
+
+                keyElement.addEventListener("click", () => {
+                    this.properties.value += this.properties.capsLock ? key.toUpperCase() : key.toLowerCase();
+                    this._triggerEvent("oninput");
+                });
+                break;
+            };
+
+            fragment.appendChild(keyElement);
+
+            if (insertLineBreak) {
+                fragment.appendChild(document.createElement("br"));
+            };
+        });
+
+        return fragment;
+    },
+
+    _triggerEvent(handlerName) {
+        if (typeof this.eventHandlers[handlerName] === "function") {
+            this.eventHandlers[handlerName](this.properties.value); 
+        };
+    },
+
+    _toggleCapsLock() {
+        this.properties.capsLock = !this.properties.capsLock;
+
+        for (let key of this.elements.keys) {
+            if (key.childElementCount === 0) {
+                key.textContent = this.properties.capsLock ? key.textContent.toUpperCase() : key.textContent.toLowerCase(); 
+            }
+        }
+    },
+
+    open(initialValue, oninput, onclose) {
+        this.properties.value = initialValue || "";
+        this.eventHandlers.oninput = oninput;
```

Comment:
Does not seem much reason to use oninput as parameter and managing collection of eventHandlers.
We can just pass into open method element which trigger event and call internal function _updateValueInTarget() where we need to update the value.

Reply:
Realised your advise. Thanks. Check, pls.